### ISSUE TITLE
[Feat]  geojson onto 3D terrain

### ISSUE THREAD

**Issue by cyyyyyyyyy** [visgl/deck.gl#6644]
### Target Use Case

I want to i using geojson/vector tile layer overlayed on generated 3D terrain.

### Proposal

![image](https://user-images.githubusercontent.com/18070879/153813891-6f221130-ffb5-4b2d-ba4c-c99696d3ba80.png)

**Comment by laijackylai** [comment]
Are there anyways to implement this with the current tool sets?

**Comment by cyyyyyyyyy** [comment]
> Are there anyways to implement this with the current tool sets?

yes, using mapbox

**Comment by laijackylai** [comment]
> yes, using mapbox
Any examples for that, cuz I can't seem to find an example for it.
Or it will just "work" when you add a geojson layer on top of the terrain layer.

**Comment by kylebarron** [comment]
I don't believe there's any support for this currently in deck.gl. As I understand it, a GPU-based solution would be pretty involved. 

As an aside, you may be interested in a [CPU-based solution for vector tiles and terrain tiles that I attempted a couple years ago](https://github.com/kylebarron/snap-to-tin).

**Comment by cyyyyyyyyy** [comment]
> > yes, using mapbox
> 
> Any examples for that, cuz I can't seem to find an example for it. Or it will just "work" when you add a geojson layer on top of the terrain layer.

sorry , what I want to express is using the mapbox-gl framework

**Comment by laijackylai** [comment]
> I don't believe there's any support for this currently in deck.gl. As I understand it, a GPU-based solution would be pretty involved.
> 
> As an aside, you may be interested in a [CPU-based solution for vector tiles and terrain tiles that I attempted a couple years ago](https://github.com/kylebarron/snap-to-tin).

Thank you @kylebarron, I'll give your library a go. Thanks a lot

**Comment by laijackylai** [comment]
> As I understand it, a GPU-based solution would be pretty involved.

As for a GPU-based solution, is it that we have to temper with the underlying WebGL code when generating the triangular mesh? Or use the generated mesh as input for getting the z-value of the vector features that will be rendered as a new layer?

**Comment by kylebarron** [comment]
> As for a GPU-based solution, is it that we have to temper with the underlying WebGL code when generating the triangular mesh?

For _generating_ the triangular mesh, that shouldn't need any changes. That's already done on the CPU via `loaders.gl/terrain` or on the backend and sent to the client.

> Or use the generated mesh as input for getting the z-value of the vector features that will be rendered as a new layer?

Yeah I think this is the harder problem. For one, for every vertex you need to find a z position, but additionally you need to interpolate between points along the terrain. For example if you have a long linestring, in order for it not to be underneath or over terrain, you'd need to create more points along the linestring for every edge of the TIN. At least this is what I tried to do in that library on CPU. Maybe there are much more efficient algorithms for GPU, but I'm not familiar with those 😄 .

**Comment by laijackylai** [comment]
thanks for the explanation. I'll try use your library to implement first 😂

**Comment by charlieforward9** [comment]
https://deck.gl/docs/api-reference/extensions/terrain-extension slightly less experimental when https://github.com/visgl/deck.gl/pull/10250 & https://github.com/visgl/deck.gl/pull/10251 land
